### PR TITLE
Task finished

### PR DESCRIPTION
**Pull request by LesterOper**
Task4 was finished
Implemented CartPageServlet, DeleteItemFromCartServlet
Implemented deleting from cart, updating cart, jsp "cartPage" and minicart.

### CODE REVIEW

File: src/main/java/com/es/phoneshop/Cart/Cart.java [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
+    public List<CartItem> getList() {
+        return cart;
+    }
+
+    public void setTotalCartCost(List<CartItem> cartItems) {
+        BigDecimal summary = new BigDecimal(0);
+        for (int i = 0; i < cartItems.size(); i++) {
+            summary = summary.add(cartItems.get(i).getTotalCost());
+        }
+        totalCartCost = summary;
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/Cart/Cart.java]: Can we use streams approach for this?

File: src/main/java/com/es/phoneshop/Cart/HttpSessionCartService.java [hunk 2]
```diff
@@ -1,72 +1,91 @@
...
+                    .findFirst()
+                    .ifPresent(item -> item.setQuantity(item.getQuantity() + quantity, product.getPrice()));
+            cartList.setTotalCartCost(cartList.getList());
+        } else if (cartList.getList().stream()
+                .noneMatch(prod -> prod.getProduct().getId().equals(product.getId()))) {
+            cartList.getList().add(new CartItem(product, quantity));
+            cartList.getList().stream().filter(prod -> prod.getProduct().getId().equals(product.getId()))
+                    .findFirst()
+                    .get().setTotalCost(product.getPrice(), quantity);
+            cartList.setTotalCartCost(cartList.getList());
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/Cart/HttpSessionCartService.java]: Please try to reuse the duplicate code.

File: src/main/java/com/es/phoneshop/web/CartPageServlet.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            return;
+        } else {
+            response.sendRedirect(request.getRequestURI() + "?message=Updated successfully");
+        }
+    }
+
+    public void showPage(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        //request.setAttribute("products",
+        //      product.getProduct(Long.parseLong(request.getRequestURI().substring(request.getRequestURI().lastIndexOf("/") + 1))));
+        request.getRequestDispatcher("/WEB-INF/pages/cartPage.jsp").
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/CartPageServlet.java]: Remove commented code.

File: src/main/java/com/es/phoneshop/web/DeleteItemFromCartServlet.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        Long productId = Long.parseLong(request.getParameter("productId"));
+        int quantity = Integer.parseInt(request.getParameter("quantity"));
+        cart = cartService.getCart(request);
+        productDao = ArrayListProductDao.getInstance();
+        Product product = productDao.getProduct(productId);
+        Iterator<CartItem> cartItem = cart.getList().iterator();
+        while (cartItem.hasNext()) {
+            CartItem item = cartItem.next();
+            if (item.getProduct().getId() == product.getId()) {
+                BigDecimal cost = item.getTotalCost();
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/DeleteItemFromCartServlet.java]: Remove unused variable.

File: src/main/webapp/WEB-INF/pages/cartPage.jsp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            <td>Description</td>
+            <td>Quantity</td>
+            <td>Price</td>
+        </tr>
+        </thead>
+        <c:set var="cart" value="${sessionScope['cartList'].list}"/>
+        <c:set var="error" value="${requestScope['errors']}"/>
+        <form id="deleteCartItem" action="${pageContext.servletContext.contextPath}/products/cart/deleteItem"
+              method="post">
+            <c:forEach items="${cart}" var="cartItem" varStatus="status">
```
Comment (SergeiStarikov) [inline src/main/webapp/WEB-INF/pages/cartPage.jsp]: What if there are no items in the cart? Will we get table header with empty rows?
It's worth checking for empty cart before rendering the table.

File: src/main/java/com/es/phoneshop/Cart/Cart.java [hunk 6]
```diff
@@ -1,7 +1,5 @@
 package com.es.phoneshop.Cart;
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/Cart/Cart.java:1]: Please add more tests for boundary cases.
